src/DatumCacheWsp.purs

```diff
@@ -0,0 +1,127 @@
+import Types.PlutusData (PlutusData, DatumHash)
+
+
+newtype BlockId = BlockId BigInt -- an integer on HS side
+
+newtype WspFault = WspFault Json
+
+faultToString :: WspFault -> String
+faultToString (WspFault j) = stringify j
+
+type JsonWspRequest =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , args :: Json
+  }
+
+type JsonWspResponse =
+  { type :: String
+  , version :: String
+  , servicename :: String
+  , methodname :: String
+  , result :: Maybe Json
+  , fault :: Maybe Json
+  }
+
+data DatumCacheRequest
+  = GetDatumByHashRequest DatumHash
+  | GetDatumsByHashesRequest (Array DatumHash)
+
+data DatumCacheResponse
+  = GetDatumByHashResponse (Maybe PlutusData)
+  | GetDatumsByHashesResponse (Array PlutusData)
+
+
+requestMethodName :: DatumCacheRequest -> String
+requestMethodName = case _ of
+  GetDatumByHashRequest _ -> "GetDatumByHash"
+  GetDatumsByHashesRequest _ -> "GetDatumsByHashes"
+
+
+jsonWspRequest :: DatumCacheRequest -> JsonWspRequest
+jsonWspRequest req =
+  { type: "jsonwsp/request"
+  , version: "1.0"
+  , servicename: "ogmios"
+  , methodname: requestMethodName req
+  , args: toArgs req
+  }
+  where
+    toArgs :: DatumCacheRequest -> Json
+    toArgs = case _ of
+      GetDatumByHashRequest dh -> encodeJson {hash: byteArrayToHex $ unwrap dh}
+      GetDatumsByHashesRequest dhs -> encodeJson {hashes: (byteArrayToHex <<< unwrap) <$> dhs}
+
+
+parseJsonWspResponse :: JsonWspResponse -> Either WspFault DatumCacheResponse
+parseJsonWspResponse resp@{methodname, result, fault} =
+  maybe
+  (Left $ maybe invalidResponseError WspFault fault)
+  decodeResponse
+  result
+  where
+    decodeResponse r = case methodname of
+      "GetDatumByHash" -> GetDatumByHashResponse <$>
+        let
+          datumFound =
+            Just <$> liftErr (decodeJson =<< deepLookup ["DatumFound", "value"] r)
+          datumNotFound =
+            Nothing <$ liftErr (deepLookup ["DatumNotFound"] r)
+        in datumFound <|> datumNotFound
+
+      "GetDatumsByHashes" -> GetDatumsByHashesResponse <$>
+          liftErr (decodeJson =<< deepLookup ["DatumFound", "value"] r)
+
+      _ -> Left invalidResponseError
+
+    liftErr :: forall a. Either JsonDecodeError a -> Either WspFault a
+    liftErr = lmap (const invalidResponseError)
+
+    invalidResponseError :: WspFault
+    invalidResponseError = WspFault $ encodeJson
+      { error: "Invalid datum cache response"
+      , response: resp}
+
+
+-- helper for assuming we get an object
+jsonObject
+  :: forall a
+   . (Object Json -> Either WspFault a)
+  -> Json
+  -> Either WspFault a
+jsonObject = caseJsonObject (Left (WspFault $ fromString "expected object"))
+
+
+
+deepLookup :: Array String -> Json -> Either JsonDecodeError Json
+deepLookup keys obj = note (UnexpectedValue obj) $ foldM lookup obj keys
+  where
+    lookup :: Json -> String -> Maybe Json
+    lookup j lbl = caseJsonObject Nothing (FO.lookup lbl) j
+
+
+-- TODO
```

Comment:
Are these TODOs for the future, or are you planning on implementing this before the PR is merged? I opened another issue for the ODC control API

Reply:
I'm fine with both approaches. What do you suggest?